File: packages/runpane-py/src/runpane/agent_context.py
from __future__ import annotations

import json
from typing import Any, Dict, List, Optional

from .generated_contract import RUNPANE_CONTRACT


def run_agent_context(parsed: Any) -> int:
    result = build_agent_context_result(parsed.context_command)
    if parsed.json:
        print(json.dumps(result, indent=2))
        return 0

    if result["mode"] == "brief":
        print(render_brief(result))
    else:
        print(render_command_detail(result["command"]))
    return 0


def build_agent_context_result(command_name: Optional[str] = None) -> Dict[str, Any]:
    if command_name:
        return {
            "ok": True,
            "mode": "command",
            "source": "runpane-contract",
            "command": get_command_detail(command_name),
        }

    brief = RUNPANE_CONTRACT["agentContext"]["brief"]
    return {
        "ok": True,
        "mode": "brief",
        "source": "runpane-contract",
        "summary": brief["summary"],
        "rules": brief["rules"],
        "tools": brief["tools"],
        "detailCommand": brief["detailCommand"],
    }


def get_command_detail(command_name: str) -> Dict[str, Any]:
    for command in RUNPANE_CONTRACT["agentContext"]["commands"].values():
        if command["name"] == command_name:
            return command

    raise ValueError(f"Unknown runpane command: {command_name}. Expected one of: {', '.join(command_names())}")


def render_brief(result: Dict[str, Any]) -> str:
    lines = [
        RUNPANE_CONTRACT["agentContext"]["brief"]["title"],
        "",
        result["summary"],
        "",
        "Rules:",
    ]
    lines.extend(f"- {rule}" for rule in result["rules"])
    lines.extend(["", "Tools:"])
    for tool in result["tools"]:
        lines.append(f"- {tool['name']}: {tool['summary']}")
        lines.append(f"  Args: {', '.join(tool['arguments'])}")
    lines.extend(["", f"Detailed definitions: {result['detailCommand']}"])
    return "\n".join(lines)


def render_command_detail(command: Dict[str, Any]) -> str:
    lines = [
        f"runpane {command['name']}",
        "",
        command["summary"],
        "",
        "Details:",
        command["details"],
        "",
        f"Requires Pane daemon: {'yes' if command.get('requiresPaneDaemon') else 'no'}",
        f"Mutates Pane state: {'yes' if command.get('mutates') else 'no'}",
        "",
        "Arguments:",
    ]

    if not command["arguments"]:
        lines.append("- none")
    else:
        for argument in command["arguments"]:
            value = f" {argument['value']}" if argument.get("value") else ""
            required = "required" if argument["required"] else "optional"
            lines.append(f"- {argument['name']}{value} ({required}): {argument['description']}")

    lines.extend(["", "Examples:"])
    lines.extend(f"- {example}" for example in command["examples"])

    if command.get("jsonSchemas"):
        lines.extend(["", "JSON schemas:"])
        lines.extend(f"- {schema}" for schema in command["jsonSchemas"])

    if command.get("notes"):
        lines.extend(["", "Notes:"])
        lines.extend(f"- {note}" for note in command["notes"])

    return "\n".join(lines)


def command_names() -> List[str]:
    return sorted(
        command["name"]
        for command in RUNPANE_CONTRACT["agentContext"]["commands"].values()
    )

File: packages/runpane/src/agentContext.ts
import { RUNPANE_CONTRACT } from './generated/contract';
import type { ParsedArgs } from './commands';

type AgentContextCommand =
  typeof RUNPANE_CONTRACT.agentContext.commands[keyof typeof RUNPANE_CONTRACT.agentContext.commands];

interface AgentContextBriefResult {
  ok: true;
  mode: 'brief';
  source: 'runpane-contract';
  summary: string;
  rules: readonly string[];
  tools: typeof RUNPANE_CONTRACT.agentContext.brief.tools;
  detailCommand: string;
}

interface AgentContextCommandResult {
  ok: true;
  mode: 'command';
  source: 'runpane-contract';
  command: AgentContextCommand;
}

type AgentContextResult = AgentContextBriefResult | AgentContextCommandResult;

export function runAgentContext(parsed: Pick<ParsedArgs, 'contextCommand' | 'json'>): number {
  const result = buildAgentContextResult(parsed.contextCommand);
  if (parsed.json) {
    console.log(JSON.stringify(result, null, 2));
    return 0;
  }

  console.log(result.mode === 'brief'
    ? renderBrief(result)
    : renderCommandDetail(result.command));
  return 0;
}

export function buildAgentContextResult(commandName?: string): AgentContextResult {
  if (commandName) {
    return {
      ok: true,
      mode: 'command',
      source: 'runpane-contract',
      command: getCommandDetail(commandName)
    };
  }

  const brief = RUNPANE_CONTRACT.agentContext.brief;
  return {
    ok: true,
    mode: 'brief',
    source: 'runpane-contract',
    summary: brief.summary,
    rules: brief.rules,
    tools: brief.tools,
    detailCommand: brief.detailCommand
  };
}

function getCommandDetail(commandName: string): AgentContextCommand {
  const normalized = normalizeCommandName(commandName);
  const detail = Object.values(RUNPANE_CONTRACT.agentContext.commands)
    .find((command) => normalizeCommandName(command.name) === normalized);
  if (detail) {
    return detail;
  }

  throw new Error(`Unknown runpane command: ${commandName}. Expected one of: ${commandNames().join(', ')}`);
}

function normalizeCommandName(commandName: string): string {
  return commandName
    .trim()
    .replace(/^runpane\s+/i, '')
    .toLowerCase()
    .replace(/[._\s-]+/g, '');
}

function renderBrief(result: AgentContextBriefResult): string {
  const lines = [
    RUNPANE_CONTRACT.agentContext.brief.title,
    '',
    result.summary,
    '',
    'Rules:',
    ...result.rules.map((rule) => `- ${rule}`),
    '',
    'Tools:',
    ...result.tools.map((tool) => `- ${tool.name}: ${tool.summary}\n  Args: ${tool.arguments.join(', ')}`),
    '',
    `Detailed definitions: ${result.detailCommand}`
  ];

  return lines.join('\n');
}

function renderCommandDetail(command: AgentContextCommand): string {
  const lines = [
    `runpane ${command.name}`,
    '',
    command.summary,
    '',
    'Details:',
    command.details,
    '',
    `Requires Pane daemon: ${command.requiresPaneDaemon ? 'yes' : 'no'}`,
    `Mutates Pane state: ${command.mutates ? 'yes' : 'no'}`,
    '',
    'Arguments:'
  ];

  if (command.arguments.length === 0) {
    lines.push('- none');
  } else {
    lines.push(...command.arguments.map((argument) => {
      const value = 'value' in argument && argument.value ? ` ${argument.value}` : '';
      const required = argument.required ? 'required' : 'optional';
      return `- ${argument.name}${value} (${required}): ${argument.description}`;
    }));
  }

  lines.push('', 'Examples:', ...command.examples.map((example) => `- ${example}`));

  const jsonSchemas = 'jsonSchemas' in command ? command.jsonSchemas : undefined;
  if (jsonSchemas?.length) {
    lines.push('', 'JSON schemas:', ...jsonSchemas.map((schema: string) => `- ${schema}`));
  }

  if (command.notes?.length) {
    lines.push('', 'Notes:', ...command.notes.map((note) => `- ${note}`));
  }

  return lines.join('\n');
}

function commandNames(): string[] {
  return Object.values(RUNPANE_CONTRACT.agentContext.commands)
    .map((command) => command.name)
    .sort();
}
